man/man8/ftpd.8
.\" Copyright (c) 1985 Regents of the University of California.
.\" All rights reserved.  The Berkeley software License Agreement
.\" specifies the terms and conditions for redistribution.
.\"
.\"	@(#)ftpd.8c	6.4 (Berkeley) 5/28/86
.\"
.TH FTPD 8
.SH NAME
ftpd, in.ftpd, setup.anonftp \- DARPA Internet File Transfer Protocol server
.SH SYNOPSIS
.B "ftp stream tcp nowait root /usr/sbin/in.ftpd in.ftpd"
.br
.B "tcpd ftp /usr/sbin/in.ftpd"
.SH DESCRIPTION
.B Ftpd
is the DARPA Internet File Transfer Prototocol
server process.  The server uses the TCP protocol
and listens at the port specified in the ``ftp''
service specification; see
.BR services (5).
.PP
The ftp server currently supports the following ftp
requests;  case is not distinguished.
.PP
.nf
.ta \w'Request        'u
\fBRequest	Description\fP
ABOR	abort previous command
ACCT	specify account (ignored)
ALLO	allocate storage (vacuously)
APPE	append to a file
CDUP	change to parent of current working directory
CWD	change working directory
DELE	delete a file
HELP	give help information
LIST	give list files in a directory (``ls -lA'')
MKD	make a directory
MODE	specify data transfer \fImode\fP
NLST	give name list of files in directory (``ls'')
NOOP	do nothing
PASS	specify password
PASV	prepare for server-to-server transfer
PORT	specify data connection port
PWD	print the current working directory
QUIT	terminate session
RETR	retrieve a file
RMD	remove a directory
RNFR	specify rename-from file name
RNTO	specify rename-to file name
STOR	store a file
STOU	store a file with a unique name
STRU	specify data transfer \fIstructure\fP
TYPE	specify data transfer \fItype\fP
USER	specify user name
XCUP	change to parent of current working directory
XCWD	change working directory
XMKD	make a directory
XPWD	print the current working directory
XRMD	remove a directory
.fi
.PP
The remaining ftp requests specified in Internet RFC 959 are
recognized, but not implemented.
.PP
The ftp server will abort an active file transfer only when the
ABOR command is preceded by a Telnet "Interrupt Process" (IP)
signal and a Telnet "Synch" signal in the command Telnet stream,
as described in Internet RFC 959.
.PP
.B Ftpd
interprets file names according to the ``globbing''
conventions used by
.BR csh (1).
This allows users to utilize the metacharacters ``*?[]{}~''.
.PP
.B Ftpd
authenticates users according to three rules. 
.IP 1)
The user name must be in the password data base,
.BR /etc/passwd ,
and not have a null password.  In this case a password
must be provided by the client before any file operations
may be performed.
.IP 2)
The user name must not appear in the file
.BR /etc/ftpusers .
.IP 3)
If the user name is ``anonymous'' or ``ftp'', an
anonymous ftp account must be present in the password
file (user ``ftp'').  In this case the user is allowed
to log in by specifying any password (by convention this
is given as the client host's name).
.PP
In the last case, 
.B ftpd
takes special measures to restrict the client's access privileges.
The server performs a 
.BR chroot (2)
command to the home directory of the ``ftp'' user.
In order that system security is not breached, it is recommended
that the ``ftp'' subtree be constructed with care;  the following
rules are recommended.
.IP ~ftp)
Make the home directory owned by ``ftp'' and unwritable by anyone.
.IP ~ftp/bin)
Make this directory owned by the super-user and unwritable by
anyone.  The program
.BR ls (1)
must be present to support the list commands.  This
program should have mode 111.
.IP ~ftp/etc)
This directory could be created, and could have
.BR passwd (5)
and
.BR group (5)
databases in it so that
.B ls
can show file ownership, but outsiders will grab your password file and
misuse it to spam you.  So don't bother.
.IP ~ftp/pub)
Make this directory mode 755 and owned by the super-user.  Create
directories in it owned by users if those users want to manage an
anonymous ftp directory.
.IP ~ftp/pub/incoming)
Optionally create this directory for anonymous uploads.  Make it mode
777.  The FTP daemon will create files with mode 266, so remote users
can write a file, but only local users can do something with it.
.PP
The script
.B setup.anonftp
can be used to create or check an anonymous FTP tree.
.SH "SEE ALSO"
.BR ftp (1).
.SH BUGS
The anonymous account is inherently dangerous and should
avoided when possible.
.ig \" MINIX 3 doesn't have privileged port numbers (yet?)
.PP
The server must run as the super-user
to create sockets with privileged port numbers.  It maintains
an effective user id of the logged in user, reverting to
the super-user only when binding addresses to sockets.  The
possible security holes have been extensively
scrutinized, but are possibly incomplete.
..


commands/ftpd200/ftpd.c
/* ftpd.c Copyright 1992-2000 by Michael Temari All Rights Reserved
 *
 * ftpd         An FTP server program for use with Minix.
 *
 * Usage:       Minix usage: tcpd ftp ftpd
 *
 * 06/14/92 Tnet Release	Michael Temari
 * 01/15/96 0.30		Michael Temari
 * 01/25/96 0.90		Michael Temari
 * 03/17/96 0.91		Michael Temari
 * 06/27/96 0.92		Michael Temari
 * 07/02/96 0.93		Michael Temari
 * 07/15/96 0.94		Michael Temari
 * 08/27/96 0.95		Michael Temari
 * 02/09/97 0.96		Michael Temari
 * 02/10/97 0.97		Michael Temari
 * 09/25/97 0.98		Michael Temari
 * 03/10/00 0.99		Michael Temari, <Michael@TemWare.Com>
 * 12/12/03 1.00		Michael Temari, <Michael@TemWare.Com>
 * 02/06/05 1.01		Michael Temari, <Michael@TemWare.Com>
 * 02/12/05 2.00		Michael Temari, <Michael@TemWare.Com>
 */

char *FtpdVersion = "2.00";

#include <sys/types.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <net/gen/in.h>
#include <net/gen/tcp.h>

#include "ftpd.h"
#include "access.h"
#include "file.h"
#include "net.h"

_PROTOTYPE(static void init, (void));
_PROTOTYPE(static int doHELP, (char *buff));
_PROTOTYPE(static int doNOOP, (char *buff));
_PROTOTYPE(static int doUNIMP, (char *buff));
_PROTOTYPE(static int getline, (char *line, int len));

FILE *msgfile = (FILE *)NULL;

/* The following defines the inactivity timeout in seconds */
#define	INACTIVITY_TIMEOUT	60*5

char *days[] = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};
char *months[] = {"Jan", "Feb", "Mar", "Apr", "May", "Jun",
		  "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"};

char line[512];

int type, format, mode, structure;
int ftpdata_fd = -1;
int loggedin, gotuser, anonymous;
char username[80];
char anonpass[128];
char newroot[128];

ipaddr_t myipaddr, rmtipaddr, dataaddr;
tcpport_t myport, rmtport, dataport;

char myhostname[256], rmthostname[256];

#define	FTPD_LOG	"/usr/adm/ftpd.log"
#define	FTPD_MSG	"/etc/ftpd_msg"

FILE *logfile;

int timeout = 0;

_PROTOTYPE(static int doHELP, (char *buff));
_PROTOTYPE(int readline, (char **args));
_PROTOTYPE(void Timeout, (int sig));
_PROTOTYPE(int main, (int argc, char *argv[]));

struct commands {
	char *name;
	_PROTOTYPE(int (*func), (char *buff));
};

struct commands commands[] = {
	"ABOR", doUNIMP,
	"ACCT", doUNIMP,
	"ALLO", doALLO,
	"APPE", doAPPE,
	"CDUP", doCDUP,
	"CWD",  doCWD,
	"DELE", doDELE,
	"HELP", doHELP,
	"LIST", doLIST,
	"MDTM", doMDTM,
	"MKD",  doMKD,
	"MODE", doMODE,
	"NLST", doNLST,
	"NOOP", doNOOP,
	"PASS", doPASS,
	"PASV", doPASV,
	"PORT", doPORT,
	"PWD",  doPWD,
	"QUIT", doQUIT,
	"REIN", doUNIMP,
	"REST", doREST,
	"RETR", doRETR,
	"RMD",  doRMD,
	"RNFR", doRNFR,
	"RNTO", doRNTO,
	"SITE", doSITE,
	"SIZE", doSIZE,
	"SMNT", doUNIMP,
	"STAT", doSTAT,
	"STOR", doSTOR,
	"STOU", doSTOU,
	"STRU", doSTRU,
	"SYST", doSYST,
	"TYPE", doTYPE,
	"USER", doUSER,
	"XCUP", doCDUP,
	"XCWD", doCWD,
	"XMKD", doMKD,
	"XPWD", doPWD,
	"XRMD", doRMD,
	"",     (int (*)())0
};

static void init()
{
   loggedin = 0;
   gotuser = 0;
   anonymous = 0;
   newroot[0] = '\0';
   type = TYPE_A;
   format = 0;
   mode = MODE_S;
   structure = 0;
   ftpdata_fd = -1;
   username[0] = '\0';
   anonpass[0] = '\0';
}

/* nothing, nada, zilch... */
static int doNOOP(buff)
char *buff;
{
   printf("200 NOOP to you too!\r\n");

   return(GOOD);
}

/* giv'em help, what a USER! */
static int doHELP(buff)
char *buff;
{
struct commands *cmd;
char star;
int i;
char *space = "    ";

   printf("214-Here is a list of available ftp commands\r\n");
   printf("    Those with '*' are not yet implemented.\r\n");

   i = 0;
   for(cmd = commands; *cmd->name != '\0'; cmd++) {
	if(cmd->func == doUNIMP)
		star = '*';
	else
		star = ' ';
	printf("     %s%c%s", cmd->name, star, space + strlen(cmd->name));
	if(++i == 6) {
		printf("\r\n");
		i = 0;
	}
   }

   if(i)
	printf("\r\n");

   printf("214 That's all the help you get.\r\n");

   return(GOOD);
}

/* not implemented */
static int doUNIMP(buff)
char *buff;
{
   printf("502 Command \"%s\" not implemented!\r\n", line);

   return(GOOD);
}

/* convert line for use */
void cvtline(args)
char **args;
{
char *p;

   p = line + strlen(line);
   while(--p >= line)
	if(*p == '\r' || *p == '\n' || isspace(*p))
		*p = '\0';
	else
		break;

  p = line;

#ifdef DEBUG
  logit("COMMAND", line);
#endif

  while(*p && !isspace(*p)) {
	*p = toupper(*p);
	p++;
  }

  if(*p) {
	*p = '\0';
	p++;
	while(*p && isspace(*p))
		p++;
   }

   *args = p;

   return;
}

static int getline(line, len)
char *line;
int len;
{
int s;
int gotcr;

   /* leave room for at end for null */
   len--;

   /* got to be able to put in at least 1 character */
   if(len < 1)
	return(-1);

   gotcr = 0;
   while(len-- > 0) {
	s = read(0, line, 1);
	if(s != 1)
		return(-1);
	if(*line == '\n')
		break;
	gotcr = (*line == '\r');
	line++;
   }
   if(gotcr)
	--line;

   *line = '\0';

   return(0);
}

int readline(args)
char **args;
{
   if(getline(line, sizeof(line)))
	return(BAD);

   cvtline(args);

   return(GOOD);
}

/* signal handler for inactivity timeout */
void Timeout(sig)
int sig;
{
   timeout = 1;

   printf("421 Inactivity timer expired.\r\n");
}

/* logit */
void logit(type, parm)
char *type;
char *parm;
{
time_t now;
struct tm *tm;

   if(logfile == (FILE *)NULL)
	return;

   time(&now);
   tm = localtime(&now);
   fprintf(logfile, "%4d%02d%02d%02d%02d%02d ",
	1900+tm->tm_year,
	tm->tm_mon + 1,
	tm->tm_mday,
	tm->tm_hour, tm->tm_min, tm->tm_sec);
   fprintf(logfile, "%s %s %s %s %s\n",
	rmthostname, username, anonymous ? anonpass : username, type, parm);
   fflush(logfile);
}

void showmsg(reply, filename)
char *reply;
char *filename;
{
FILE *mfp;
char *pe;
static char mline[256];

   if(filename == (char *)NULL)
	mfp = msgfile;
   else
	mfp = fopen(filename, "r");

   if(mfp == (FILE *)NULL)
	return;

   while(fgets(mline, sizeof(mline), mfp) != (char *)NULL) {
	pe = mline + strlen(mline);
	while(--pe >= mline)
		if(*pe == '\r' || *pe == '\n')
			*pe = '\0';
		else
			break;
	printf("%s- %s\r\n", reply, mline);
   }

   if(filename != (char *)NULL)
	fclose(mfp);
}

int main(argc, argv)
int argc;
char *argv[];
{
struct commands *cmd;
char *args;
int status;
time_t now;
struct tm *tm;
int s;

   GetNetInfo();

   /* open transfer log file if it exists */
   if((logfile = fopen(FTPD_LOG, "r")) != (FILE *)NULL) {
	fclose(logfile);
	logfile = fopen(FTPD_LOG, "a");
   }

   /* open login msg file */
   msgfile = fopen(FTPD_MSG, "r");

   /* Let's initialize some stuff */
   init();

   /* Log the connection */
   logit("CONNECT", "");

   /* Tell 'em we are ready */
   time(&now);
   tm = localtime(&now);
   printf("220 FTP service (Ftpd %s) ready on %s at ",
   	FtpdVersion, myhostname);
   printf("%s, %02d %s %d %02d:%02d:%02d %s\r\n", days[tm->tm_wday],
   	tm->tm_mday, months[tm->tm_mon], 1900+tm->tm_year,
	tm->tm_hour, tm->tm_min, tm->tm_sec,
	tzname[tm->tm_isdst]);
   fflush(stdout);

   /* Loop here getting commands */
   while(1) {
	signal(SIGALRM, Timeout);
	alarm(INACTIVITY_TIMEOUT);
	if(readline(&args) != GOOD) {
		if(!timeout)
			printf("221 Control connection closing (EOF).\r\n");
		break;
	}
	alarm(0);
	for(cmd = commands; *cmd->name != '\0'; cmd++)
		if(!strcmp(line, cmd->name))
			break;
	if(*cmd->name != '\0')
		status = (*cmd->func)(args);
	else {
		printf("500 Command \"%s\" not recognized.\r\n", line);
		status = GOOD;
	}
	fflush(stdout);
	if(status != GOOD)
		break;
   }

   CleanUpPasv();

   return(-1);
}
